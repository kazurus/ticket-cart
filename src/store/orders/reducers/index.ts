import { OrdersActionTypes } from '../actionTypes';
import type { OrdersActions } from '../actionTypes';
import type { OrderId, Order } from '../types';

export type OrdersMapById = Record<OrderId, Order>;

export type OrdersState = {
  entities: {
    byId: OrdersMapById;
    allIds: OrderId[];
  };
};

export const initialState: OrdersState = {
  entities: {
    byId: {},
    allIds: [],
  },
};

// Simple id generation
let nextOrderId = 1;

export function ordersReducer(
  state = initialState,
  action: OrdersActions
): OrdersState {
  switch (action.type) {
    case OrdersActionTypes.ADD_ORDER: {
      const { ticketId } = action.payload;
      const order: Order = {
        // eslint-disable-next-line no-plusplus
        id: nextOrderId++,
        ticketId,
        createdAt: Date.now(),
      };
      const byId = { ...state.entities.byId, [order.id]: order };
      const allIds = state.entities.allIds.concat(order.id);

      return {
        ...state,
        entities: {
          byId,
          allIds,
        },
      };
    }

    case OrdersActionTypes.DELETE_ORDER: {
      const { orderId } = action.payload;
      const { [orderId]: value, ...byId } = state.entities.byId;
      const allIds = state.entities.allIds.filter((id) => id !== orderId);

      return {
        ...state,
        entities: {
          byId,
          allIds,
        },
      };
    }

    default:
      return state;
  }
}
