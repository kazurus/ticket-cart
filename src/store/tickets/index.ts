export { ticketsActions } from './actions';
export { watchLoadTickets } from './sagas';
export { ticketsReducer } from './reducers';
export type { TicketsState } from './reducers';
