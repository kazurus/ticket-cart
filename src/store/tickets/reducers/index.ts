import { TicketsActionTypes, TicketsLoadError } from '../actionTypes';
import type { TicketsAction } from '../actionTypes';
import { TicketId, Ticket } from '../types';

export enum TicketsLoadStatus {
  IDLE = 'IDLE',
  PENDING = 'PENDING',
  READY = 'READY',
  ERROR = 'ERROR',
}

export type TicketsMapById = Record<TicketId, Ticket>;

export type TicketsState = {
  ticketsLoadStatus: TicketsLoadStatus;
  error: TicketsLoadError;
  entities: {
    byId: TicketsMapById;
    allIds: TicketId[];
  };
};

export const initialState: TicketsState = {
  ticketsLoadStatus: TicketsLoadStatus.IDLE,
  error: '',
  entities: {
    byId: {},
    allIds: [],
  },
};

function loadTicketsReducer(
  state: TicketsState,
  action: TicketsAction
): TicketsState {
  switch (action.type) {
    case TicketsActionTypes.FETCH_TICKETS: {
      return {
        ...state,
        ticketsLoadStatus: TicketsLoadStatus.PENDING,
        error: '',
      };
    }

    default:
      return state;
  }
}

function loadTicketsPendingReducer(
  state: TicketsState,
  action: TicketsAction
): TicketsState {
  switch (action.type) {
    case TicketsActionTypes.FETCH_TICKETS_SUCCESS: {
      const { tickets } = action.payload;
      const ticketsMapById = tickets.reduce(
        (ticketsMap, ticket) => ({
          ...ticketsMap,
          [ticket.id]: ticket,
        }),
        {}
      );
      const ticketsIds = tickets.map((ticket) => ticket.id);

      return {
        ...state,
        ticketsLoadStatus: TicketsLoadStatus.READY,
        entities: {
          byId: ticketsMapById,
          allIds: ticketsIds,
        },
      };
    }

    case TicketsActionTypes.FETCH_TICKETS_FAILURE: {
      return {
        ...state,
        ticketsLoadStatus: TicketsLoadStatus.ERROR,
        error: action.payload.error,
      };
    }

    default:
      return state;
  }
}

export function ticketsReducer(
  state: TicketsState = initialState,
  action: TicketsAction
): TicketsState {
  switch (state.ticketsLoadStatus) {
    case TicketsLoadStatus.IDLE:
    case TicketsLoadStatus.ERROR:
      return loadTicketsReducer(state, action);
    case TicketsLoadStatus.PENDING:
      return loadTicketsPendingReducer(state, action);

    default:
      return state;
  }
}
