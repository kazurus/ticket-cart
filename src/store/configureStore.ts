import createSagaMiddleware from 'redux-saga';
import { createStore, applyMiddleware, Middleware, compose } from 'redux';
import { createLogger } from 'redux-logger';

import { rootReducer } from './reducers';
import { rootSaga } from './sagas';

export function configureAppStore() {
  const sagaMiddleware = createSagaMiddleware();

  const middlewares: Middleware[] = [sagaMiddleware];

  const excludeEnv = ['production', 'test'];

  if (!excludeEnv.includes(process.env.NODE_ENV)) {
    middlewares.push(createLogger());
  }

  const composeEnhancer =
    (process.env.NODE_ENV !== 'production' &&
      ((window as any)
        .__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ as typeof compose)) ||
    compose;

  const enhancer = composeEnhancer(applyMiddleware(...middlewares));

  const store = createStore(rootReducer, enhancer);

  sagaMiddleware.run(rootSaga);

  return store;
}
