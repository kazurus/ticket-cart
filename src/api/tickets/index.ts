import { mapDestinationsToTickets } from './mappers';
import { destinationsListMock, milePriceMock } from './mockData';
import type { Ticket } from '../../store/tickets/types';

export const getTickets: () => Promise<Ticket[]> = async () =>
  mapDestinationsToTickets(destinationsListMock, milePriceMock);

const tickets = {
  getTickets,
};

export default tickets;
