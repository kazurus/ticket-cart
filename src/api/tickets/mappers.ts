import { TicketId, Ticket } from '../../store/tickets/types';
import type { Destination, MilePrice } from './mockData';

export type TicketInfo = {
  id: TicketId;
  from: Destination;
  to: Destination;
};

/**
 * Add price to tickets, the price is calculated by Taxicab geometry
 * https://en.wikipedia.org/wiki/Taxicab_geometry
 */
const mapTicketsInfoToTickets = (
  ticketsInfo: TicketInfo[],
  milePrice: MilePrice
) =>
  ticketsInfo.map((direction) => {
    const { from, to } = direction;
    const { coordinates: fromCoordinates } = from;
    const { coordinates: toCoordinates } = to;

    const roadLengthByX = fromCoordinates.x - toCoordinates.x;
    const roadLengthByY = fromCoordinates.y - toCoordinates.y;
    const roadLength = Math.abs(roadLengthByX) + Math.abs(roadLengthByY);
    const roadPrice = roadLength * milePrice;

    const ticket: Ticket = {
      ...direction,
      from: direction.from.title,
      to: direction.to.title,
      price: roadPrice,
    };

    return ticket;
  });

export const mapDestinationsToTickets = (
  destinations: Destination[],
  milePrice: MilePrice
) => {
  // Simple generation for ticket ids
  let nextId = 1;

  const allTicketsInfo: TicketInfo[] = destinations.reduce(
    (acc, currentDestination, index) => {
      const ticketsInfo: TicketInfo[] = destinations
        .slice(index + 1)
        .map((nextDestination) => ({
          /* eslint-disable-next-line no-plusplus */
          id: nextId++,
          from: { ...currentDestination },
          to: { ...nextDestination },
        }))
        .concat(acc);

      return ticketsInfo;
    },
    [] as TicketInfo[]
  );

  return mapTicketsInfoToTickets(allTicketsInfo, milePrice);
};
