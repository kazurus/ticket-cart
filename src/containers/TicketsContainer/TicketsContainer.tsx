import React from 'react';
import { useSelector } from 'react-redux';
import TicketItem from '../../components/tickets/TicketItem';
import TicketsList from '../../components/tickets/TicketsList';
import { getTickets } from '../../store/tickets/selectors';
import type { RootState } from '../../store/reducers';

const TicketsContainer: React.FC = () => {
  const tickets = useSelector((state: RootState) => getTickets(state));

  return (
    <TicketsList>
      {tickets.map((ticket) => (
        <TicketItem key={ticket.id} ticket={ticket} />
      ))}
    </TicketsList>
  );
};

export default TicketsContainer;
