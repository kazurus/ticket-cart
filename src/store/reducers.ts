import { combineReducers } from 'redux';
import { ticketsReducer } from './tickets';
import type { TicketsState } from './tickets';

export type RootState = {
  tickets: TicketsState;
};

export const rootReducer = combineReducers({
  tickets: ticketsReducer,
});
