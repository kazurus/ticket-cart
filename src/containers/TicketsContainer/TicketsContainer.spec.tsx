import React from 'react';
import { Store } from 'redux';
import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import { configureAppStore } from '../../store/configureStore';
import { ticketsActions } from '../../store/tickets';
import { initialState } from '../../store/tickets/reducers';
import TicketsContainer from './TicketsContainer';

const renderOrdersContainer = (store: Store) =>
  render(
    <Provider store={store}>
      <TicketsContainer />
    </Provider>
  );

jest.mock('redux-saga/effects', () => {
  const operators = jest.requireActual('redux-saga/effects');
  const delay = (s: number) => new Promise((res) => setTimeout(res, 0));
  return { ...operators, delay };
});

describe('TicketsContainer component', () => {
  let store: ReturnType<typeof configureAppStore>;
  let component: ReturnType<typeof renderOrdersContainer>;

  beforeEach(() => {
    store = configureAppStore();
    component = renderOrdersContainer(store);
    expect(store.getState().tickets).toEqual(initialState);
    store.dispatch(ticketsActions.loadTickets());
  });

  afterEach(() => {
    component.unmount();
  });

  it('should render components', () => {
    const { container } = component;
    expect(container).toMatchSnapshot();
  });

  it('should have empty list of tickets on mount', () => {
    const { queryAllByTestId } = component;
    expect(queryAllByTestId('ticket-item')).toHaveLength(0);
  });

  it('should render list of tickets', async () => {
    const { findAllByTestId } = component;
    const tickets = await findAllByTestId('ticket-item');

    expect(tickets.length).toBeGreaterThan(2);
  });
});
