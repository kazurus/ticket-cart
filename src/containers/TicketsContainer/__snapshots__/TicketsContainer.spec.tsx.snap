// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TicketsContainer component should render components 1`] = `
<div>
  <div>
    <h3>
      Tickets
    </h3>
    <div />
  </div>
</div>
`;
